packages/pool/src/weighted.rs

```diff
@@ -0,0 +1,407 @@
+            amount_in,
+            output_amount.amount.into(),
+        )?;
+
+        Ok(amount_in)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::Pool;
+    use crate::{test::load_pools, OsmosisPool};
+
+    use std::{collections::BTreeMap, str::FromStr};
+
+    use anyhow::anyhow;
+    use cosmwasm_std::{coin, testing::mock_dependencies, Coin, Deps, Uint256};
+
+    fn calc_out(
+        deps: Deps,
+        pools: &mut BTreeMap<u64, Pool>,
+        pool_id: u64,
+        input: Coin,
+        output: &str,
+    ) -> anyhow::Result<Uint256> {
+        if let Pool::Weighted(pool) = pools.get_mut(&pool_id).unwrap() {
+            let amount_out = pool.swap_exact_amount_in(
+                &deps,
+                input,
+                output.to_string(),
+                Uint256::from_str("100")?,
+                pool.get_spread_factor()?,
+            )?;
+
+            Ok(amount_out)
+        } else {
+            Err(anyhow!("pool type is not weighted"))
+        }
+    }
+
+    fn calc_in(
+        deps: Deps,
+        pools: &mut BTreeMap<u64, Pool>,
+        pool_id: u64,
+        input: &str,
+        output: Coin,
+    ) -> anyhow::Result<Uint256> {
+        if let Pool::Weighted(pool) = pools.get_mut(&pool_id).unwrap() {
+            let amount_in = pool.swap_exact_amount_out(
+                &deps,
+                input.to_string(),
+                Uint256::from_str("100")?,
+                output,
+                pool.get_spread_factor()?,
+            )?;
+
+            Ok(amount_in)
+        } else {
+            Err(anyhow!("pool type is not weighted"))
+        }
+    }
+
+    #[test]
+    fn test_sim_in() -> anyhow::Result<()> {
+        let deps = mock_dependencies();
+        let mut pools = load_pools("./tests/testdata/all-pools-after.json".into())?;
+
+        let cases = [
+            (
+                1,
+                coin(100_000_000_000, "uosmo"),
+                "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
+            ),
+            (
+                722,
+                coin(100_000_000_000, "uosmo"),
+                "ibc/6AE98883D4D5D5FF9E50D7130F1305DA2FFA0C652D1DD9C123657C6B4EB2DF8A",
+            ),
+            (
+                584,
+                coin(100_000_000_000, "uosmo"),
+                "ibc/0954E1C28EB7AF5B72D24F3BC2B47BBB2FDF91BDDFD57B74B99E133AED40972A",
+            ),
+            (2, coin(100_000_000, "uosmo"), "uion"),
+            (
+                3,
+                coin(
+                    4_946_633,
+                    "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4",
+                ),
+                "uosmo",
+            ),
+        ];
+
+        for (pool_id, input, output) in cases {
+            println!("Trying: {} -> {}", input, output);
+            let res = calc_out(deps.as_ref(), &mut pools, pool_id, input.clone(), output)?;
+            println!("=> {}{}\n", res, output);
```

Comment:
you should check correctness by eye?

Reply:
I'll change it to compare the result with test-tube returns